### PR TITLE
[WIP] Private repo

### PR DESCRIPTION
**Pull request by octavioamu**
##### Checklist

- [x] Read and conforms to the [contributor guidelines](https://docs.gitcoin.co/mk_contributors/)
- [x] Read and conforms to the [style guidelines](https://docs.gitcoin.co/mk_styleguide/)
- [x] Linter status: 100% pass
- [x] Added relevant tests
- [x] Tests pass and code coverage has not decreased
- [x] Changes don't break existing behavior
- [x] Commit message follows [commit guidelines](https://docs.gitcoin.co/mk_contributors/#step-4-commit)
- [x] [Code documentation](https://docs.gitcoin.co/mk_contributors/#docstrings)
- [x] Completed: Code review by core team and any requested changes

##### Refers/Fixes
Ref #3675

###### Contributor

- [ ] Read and followed the [Contributor Guidelines](https://docs.gitcoin.co/mk_contributors/)
- [ ] Tested all changes **locally**
- [ ] Verified existing functionality
- [ ] Ran `make test` and everything passed!

###### Reviewer

- [ ] Affirm contributor guidelines have been followed and requested changes made
- [ ] CI tests and linting pass
- [ ] No conflicts (migrations, files, etc)
- [ ] Regression tested against staging or local deployment

###### Funder

- [ ] Validated requested changes were made to specification
- [ ] Bounty payout released to the contributor

### CODE REVIEW

**Comment by danlipert** [comment]
@octavioamu is this ready for review or still a WIP?

**Comment by octavioamu** [comment]
@danlipert I think is read for review, I will fix the conflict

File: app/app/utils.py [hunk 1]
```diff
@@ -161,6 +162,7 @@ def setup_lang(request, user):
         DoesNotExist: The exception is raised if no profile is found for the specified handle.
 
     """
+    from dashboard.models import Profile
```
Comment (danlipert) [inline app/app/utils.py]: Why did you move this import within the function from the top of the file?

File: app/assets/v2/js/pages/bounty_details.js [hunk 2]
```diff
@@ -633,51 +633,181 @@ var attach_override_status = function() {
...
+              signed_nda: response.bounty_doc_id
+            }).then(success => {
+              if (success) {
+                $(self).attr('href', '/uninterested');
+                $(self).find('span').text(gettext('Stop Work'));
+                $(self).parent().attr('title', '<div class="tooltip-info tooltip-sm">' + gettext('Notify the funder that you will not be working on this project') + '</div>');
+                $.modal.close();
+              }
+            }).catch((error) => {
+              if (error.responseJSON.error === 'You may only work on max of 3 issues at once.')
```
Comment (danlipert) [inline app/assets/v2/js/pages/bounty_details.js]: Is there a better way we can check for this error? I worry that in the future we may change this error message and not realize that it will break this.

File: app/assets/v2/js/pages/bounty_details.js [hunk 3]
```diff
@@ -633,51 +633,181 @@ var attach_override_status = function() {
...
-          throw error;
-        });
       });
+
     });
   });
+  modals.bootstrapModal('show');
+
+
+  // setTimeout(function() {
```
Comment (danlipert) [inline app/assets/v2/js/pages/bounty_details.js]: Should we remove this commented out function?

File: app/assets/v2/js/pages/bounty_details.js [hunk 4]
```diff
@@ -1266,6 +1397,10 @@ var pull_bounty_from_api = function() {
         render_activity(result, results);
 
         document.result = result;
+
+        if (typeof promptPrivateInstructions != 'undefined' && result.repo_type === 'private') {
```
Comment (danlipert) [inline app/assets/v2/js/pages/bounty_details.js]: not a big deal, but should we use `!==` instead of `!=` here?

File: app/assets/v2/js/pages/new_bounty.js [hunk 5]
```diff
@@ -144,10 +155,10 @@ $(document).ready(function() {
     $('input[name=revisions]').val(revision);
   });
 
-  if ($('input[name=issueURL]').val() != '') {
+  if ($('input[name=issueURL]').val() != '' && !isPrivateRepo) {
     retrieveIssueDetails();
   }
-  $('input[name=issueURL]').focus();
+  // $('input[name=issueURL]').focus();
```
Comment (danlipert) [inline app/assets/v2/js/pages/new_bounty.js]: remove comment here

File: app/assets/v2/js/pages/new_bounty.js [hunk 6]
```diff
@@ -101,7 +111,8 @@ $(document).ready(function() {
   $('input[name=hours]').blur(setUsdAmount);
   $('select[name=denomination]').change(setUsdAmount);
   $('select[name=denomination]').change(promptForAuth);
-  $('input[name=issueURL]').blur(retrieveIssueDetails);
+
+  // !isPrivateRepo ? $('input[name=issueURL]').blur(retrieveIssueDetails) : undefined;
```
Comment (danlipert) [inline app/assets/v2/js/pages/new_bounty.js]: remove comment here

File: app/assets/v2/js/pages/new_bounty.js [hunk 7]
```diff
@@ -510,3 +562,70 @@ getAmountEstimate(usdFeaturedPrice, 'ETH', (amountEstimate) => {
   ethFeaturedPrice = amountEstimate['value'];
   $('.featured-price-eth').text(`+${amountEstimate['value']} ETH`);
 });
+
+
+let isPrivateRepo = false;
+let params = (new URL(document.location)).searchParams;
+
+const setPrivateForm = () => {
+  // $('#description')
```
Comment (danlipert) [inline app/assets/v2/js/pages/new_bounty.js]: extra comments here

File: app/dashboard/helpers.py [hunk 8]
```diff
@@ -433,6 +445,8 @@ def create_new_bounty(old_bounties, bounty_payload, bounty_details, bounty_id):
                 'featuring_date': timezone.make_aware(
                     timezone.datetime.fromtimestamp(metadata.get('featuring_date', 0)),
                     timezone=UTC),
+                'repo_type': metadata.get('repo_type', None),
+                'unsigned_nda': unsigned_nda if unsigned_nda else None,
```
Comment (danlipert) [inline app/dashboard/helpers.py]: Do we need to check for `unsigned_nda` here if its already set to `None`? Could it be something other than a `BountyDocuments` object or `None`?

File: app/dashboard/models.py [hunk 9]
```diff
@@ -271,10 +275,12 @@ class Bounty(SuperModel):
     canceled_bounty_reason = models.TextField(default='', blank=True, verbose_name=_('Cancelation reason'))
     project_type = models.CharField(max_length=50, choices=PROJECT_TYPES, default='traditional')
     permission_type = models.CharField(max_length=50, choices=PERMISSION_TYPES, default='permissionless')
+    repo_type = models.CharField(max_length=50, choices=REPO_TYPES, default='public')
     snooze_warnings_for_days = models.IntegerField(default=0)
     is_featured = models.BooleanField(
         default=False, help_text=_('Whether this bounty is featured'))
     featuring_date = models.DateTimeField(blank=True, null=True)
+    unsigned_nda = models.ForeignKey('dashboard.BountyDocuments', blank=True, null=True, related_name='unsignednda', on_delete=models.SET_NULL)
```
Comment (danlipert) [inline app/dashboard/models.py]: The `related_name` here is confusing, since it refers to a `Bounty` object - how about `bounty`?

File: app/dashboard/models.py [hunk 10]
```diff
@@ -1457,6 +1469,7 @@ class Interest(SuperModel):
         max_length=7,
         help_text=_('Whether or not the interest requires review'),
         verbose_name=_('Needs Review'))
+    signed_nda = models.ForeignKey('dashboard.BountyDocuments', blank=True, null=True, related_name='signednda', on_delete=models.SET_NULL)
```
Comment (danlipert) [inline app/dashboard/models.py]: Same comment about the related_name for Bounties above

File: app/app/utils.py (hunk 1 above)
Comment (SaptakS) [reply app/app/utils.py]: Aah. i need to remove this. I was previously importing util in dashboard and it was causing a recursive importing, so. Thanks for pointing out.

Comment (danlipert) [reply app/app/utils.py]: How much time do you think it would take to refactor out the recursive import? Imports within a function aren't terribly bad, but I think for webservers they can really affect performance.

**Comment by danlipert** [comment]
@octavioamu @SaptakS Looks like a lot of great fixes! Just one comment and then this branch needs a rebase. Also, can someone fill out the checkboxes in the PR?

File: app/app/utils.py (hunk 1 above)
Comment (SaptakS) [reply app/app/utils.py]: It will take a lot of adjustments in a lot of places. If it's not too much bad for the performance, I think we can do it in a later adjustment.

**Comment by SaptakS** [comment]
Fixed merge conflicts.
